### PR TITLE
Add CoreOS reboot via reboot-api user.

### PR DESCRIPTION
**Pull request by robertodauria**
This PR adds the "host" reboot method to the `/v1/reboot` endpoint. Such method will execute a `systemctl reboot` on the CoreOS host by logging in via SSH with a previously defined `reboot-api` user and a private key. 

New metrics:
- `reboot_coreos_total{site, machine}`: number of reboots executed via the "host" method for a given node.

---
This change is [Reviewable](https://reviewable.io/reviews/m-lab/reboot-service/6)

### CODE REVIEW

**Review by stephen-soltesz — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/m-lab/reboot-service/6)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @evfirerob and @pboothe)

___
*[main.go, line 27 at r1](https://reviewable.io/reviews/m-lab/reboot-service/6#-Lbi-eFO1wlC0_Go5_gK:-Lbi-eFO1wlC0_Go5_gL:b-rkc960) ([raw file](https://github.com/m-lab/reboot-service/blob/ee45aff7246c7cbf1eb86f336417d122da4da414/main.go#L27)):*
> ```Go
> 	promAddr   = flag.String("promaddr", defaultPromPort,
> 		"Address to listen on for Prometheus metrics")
> 	projectID  = flag.String("datastore.project", defaultProjID, "GCD project ID")
> ```

I really like these namespace'd flag names!
___
*[main.go, line 81 at r1](https://reviewable.io/reviews/m-lab/reboot-service/6#-Lbi2HeR6NJ1pPQ9zuCF:-Lbi2HeR6NJ1pPQ9zuCG:b-ku04b9) ([raw file](https://github.com/m-lab/reboot-service/blob/ee45aff7246c7cbf1eb86f336417d122da4da414/main.go#L81)):*
> ```Go
> 	// Initialize HTTP server.
> 	rebootMux := http.NewServeMux()
> 	rebootMux.Handle("/v1/reboot", rebootHandler)
> ```

In this or a future PR, please consider adding `promhttp` instruments for handlers. e.g. like https://github.com/m-lab/ndt-server/blob/master/ndt-server.go#L121
___
*[reboot/handler.go, line 35 at r1](https://reviewable.io/reviews/m-lab/reboot-service/6#-Lbi03-ZELtCnykR3Etx:-Lbi03-ZELtCnykR3Ety:b-fhdpen) ([raw file](https://github.com/m-lab/reboot-service/blob/ee45aff7246c7cbf1eb86f336417d122da4da414/reboot/handler.go#L35)):*
> ```Go
> 	metricCoreOSReboots = promauto.NewCounterVec(
> 		prometheus.CounterOpts{
> 			Name: "reboot_coreos_total",
> ```

Rather than 'coreos', I might suggest a technology independent name like 'host'? or similar. "host" matches the "method" name, which is nice.

**Review by robertodauria — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/m-lab/reboot-service/6)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @pboothe and @stephen-soltesz)

___
*[main.go, line 81 at r1](https://reviewable.io/reviews/m-lab/reboot-service/6#-Lbi2HeR6NJ1pPQ9zuCF:-LbxYgQy6Dwlb3M_1UqR:b-37l3us) ([raw file](https://github.com/m-lab/reboot-service/blob/ee45aff7246c7cbf1eb86f336417d122da4da414/main.go#L81)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

In this or a future PR, please consider adding `promhttp` instruments for handlers. e.g. like https://github.com/m-lab/ndt-server/blob/master/ndt-server.go#L121
</blockquote></details>

I will but I'd like to keep it on a dedicated PR.
___
*[reboot/handler.go, line 35 at r1](https://reviewable.io/reviews/m-lab/reboot-service/6#-Lbi03-ZELtCnykR3Etx:-LbxYoxjErrS_-2q3VRx:b7tw0dq) ([raw file](https://github.com/m-lab/reboot-service/blob/ee45aff7246c7cbf1eb86f336417d122da4da414/reboot/handler.go#L35)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Rather than 'coreos', I might suggest a technology independent name like 'host'? or similar. "host" matches the "method" name, which is nice.
</blockquote></details>

Done. I've left a single reference in the help string for the `reboot.user` flag to clarify which kind of hosts it refers to (just "for rebooting hosts" seemed a bit ambiguous to me in a place where it can't be inferred from the context).

**Comment by pboothe** [comment]
Looks like you and Stephen are doing good things here. I am removing myself from the approvers list.

**Review by stephen-soltesz — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/m-lab/reboot-service/6)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @evfirerob and @stephen-soltesz)

___
*[main.go, line 81 at r1](https://reviewable.io/reviews/m-lab/reboot-service/6#-Lbi2HeR6NJ1pPQ9zuCF:-Lbxd4UUDcplSyhH8dMu:b2od5qx) ([raw file](https://github.com/m-lab/reboot-service/blob/ee45aff7246c7cbf1eb86f336417d122da4da414/main.go#L81)):*
<details><summary><i>Previously, evfirerob (Roberto D'Auria) wrote…</i></summary><blockquote>

I will but I'd like to keep it on a dedicated PR.
</blockquote></details>

Please add a TODO.
___
*[connector/connector.go, line 142 at r2](https://reviewable.io/reviews/m-lab/reboot-service/6#-Lbxdee91d3icPByZuw6:-Lbxdee91d3icPByZuw7:bm8py8h) ([raw file](https://github.com/m-lab/reboot-service/blob/641b0d525d231f3893cc2a18f2539f2e14eeee88/connector/connector.go#L142)):*
> ```Go
> 			return "", err
> 		}
> 	} else { // reboot via DRAC (default)
> ```

Please check explicitly for `BMCConnection`.
___
*[reboot/handler.go, line 76 at r2](https://reviewable.io/reviews/m-lab/reboot-service/6#-LbyUKPND1WzpiOIwm9I:-LbyUKPND1WzpiOIwm9J:b2gdbl6) ([raw file](https://github.com/m-lab/reboot-service/blob/641b0d525d231f3893cc2a18f2539f2e14eeee88/reboot/handler.go#L76)):*
> ```Go
> }
> 
> func (h *Handler) rebootHost(ctx context.Context, node string, site string) (string, error) {
> ```

Overall coverage has dropped significantly in this PR -- is it possible to test the new paths?
___
*[reboot/handler.go, line 79 at r2](https://reviewable.io/reviews/m-lab/reboot-service/6#-LbyQRlf601ClAWYj1IB:-LbyQRlf601ClAWYj1IC:b-vapx9) ([raw file](https://github.com/m-lab/reboot-service/blob/641b0d525d231f3893cc2a18f2539f2e14eeee88/reboot/handler.go#L79)):*
> ```Go
> 	// To reboot a host a "reboot-api" user is created, and the only way
> 	// to authenticate is via a private key. Logging in with such user will
> 	// automatically trigger a "systemctl reboot" command.
> ```

This comment would be more helpful near the `exec("")` command in connector.go which looks odd by itself, but makes more sense if we know that the reboot-api user behaves in the way described here.
___
*[reboot/handler.go, line 105 at r2](https://reviewable.io/reviews/m-lab/reboot-service/6#-LbyQrSo2NsE4eDbguPr:-LbyQrSo2NsE4eDbguPs:b-iguq8x) ([raw file](https://github.com/m-lab/reboot-service/blob/641b0d525d231f3893cc2a18f2539f2e14eeee88/reboot/handler.go#L105)):*
> ```Go
> 	}
> 
> 	metricHostReboots.WithLabelValues(site, node).Inc()
> ```

What do you think of also counting the errors? This could be another label for `status={ok, error}`. distinguishing the errors might also be helpful. e.g. error-connect, error-reboot, or similar.  If these errors propagate all the way back to the http request, then `promhttp` instruments might be enough too. I'll let you decide.

**Review by stephen-soltesz — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/m-lab/reboot-service/6)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @evfirerob)

___
*[main.go, line 25 at r2](https://reviewable.io/reviews/m-lab/reboot-service/6#-Lbyovv7Eyf8DHsX0mle:-Lbyovv7Eyf8DHsX0mlf:bg6z08n) ([raw file](https://github.com/m-lab/reboot-service/blob/641b0d525d231f3893cc2a18f2539f2e14eeee88/main.go#L25)):*
> ```Go
> 	// Command line flags.
> 	listenAddr = flag.String("listenaddr", defaultListenAddr, "Address to listen on")
> 	promAddr   = flag.String("promaddr", defaultPromPort,
> ```

This function (just added) https://godoc.org/github.com/m-lab/go/prometheusx#MustServeMetrics

uses a standard package flag for the prometheus metrics server. Please use this instead of `prometheusx.MustStartPrometheus(*promAddr)`.

**Review by robertodauria — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/m-lab/reboot-service/6)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @evfirerob and @stephen-soltesz)

___
*[connector/connector.go, line 142 at r2](https://reviewable.io/reviews/m-lab/reboot-service/6#-Lbxdee91d3icPByZuw6:-Lc1847G50NXuNgVWzcG:b-896fix) ([raw file](https://github.com/m-lab/reboot-service/blob/641b0d525d231f3893cc2a18f2539f2e14eeee88/connector/connector.go#L142)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Please check explicitly for `BMCConnection`.
</blockquote></details>

Done.
___
*[reboot/handler.go, line 79 at r2](https://reviewable.io/reviews/m-lab/reboot-service/6#-LbyQRlf601ClAWYj1IB:-Lc15Fh2Cr2n0KRHgOAi:b-2k123n) ([raw file](https://github.com/m-lab/reboot-service/blob/641b0d525d231f3893cc2a18f2539f2e14eeee88/reboot/handler.go#L79)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

This comment would be more helpful near the `exec("")` command in connector.go which looks odd by itself, but makes more sense if we know that the reboot-api user behaves in the way described here.
</blockquote></details>

Done. True, thanks!

**Review by robertodauria — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/m-lab/reboot-service/6)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @evfirerob and @stephen-soltesz)

___
*[main.go, line 81 at r1](https://reviewable.io/reviews/m-lab/reboot-service/6#-Lbi2HeR6NJ1pPQ9zuCF:-Lc1AMZD6Az9lwm5cXPF:b-896fix) ([raw file](https://github.com/m-lab/reboot-service/blob/ee45aff7246c7cbf1eb86f336417d122da4da414/main.go#L81)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Please add a TODO.
</blockquote></details>

Done.
___
*[reboot/handler.go, line 105 at r2](https://reviewable.io/reviews/m-lab/reboot-service/6#-LbyQrSo2NsE4eDbguPr:-Lc1EMkp9zpVwgtY5JCY:b-mnkrd9) ([raw file](https://github.com/m-lab/reboot-service/blob/641b0d525d231f3893cc2a18f2539f2e14eeee88/reboot/handler.go#L105)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

What do you think of also counting the errors? This could be another label for `status={ok, error}`. distinguishing the errors might also be helpful. e.g. error-connect, error-reboot, or similar.  If these errors propagate all the way back to the http request, then `promhttp` instruments might be enough too. I'll let you decide.
</blockquote></details>

Done. What I generally do is to propagate the error back to the handler and return an HTTP error code that makes sense, but I agree that having more granularity and reporting the error's cause in a metric is much more useful.

**Review by robertodauria — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/m-lab/reboot-service/6)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @evfirerob and @stephen-soltesz)

___
*[reboot/handler.go, line 76 at r2](https://reviewable.io/reviews/m-lab/reboot-service/6#-LbyUKPND1WzpiOIwm9I:-Lc1Q4wG6xq6DLyI1C-Z:b-896fix) ([raw file](https://github.com/m-lab/reboot-service/blob/641b0d525d231f3893cc2a18f2539f2e14eeee88/reboot/handler.go#L76)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Overall coverage has dropped significantly in this PR -- is it possible to test the new paths?
</blockquote></details>

Done.

**Review by stephen-soltesz — APPROVED** [review]
:lgtm:

> Reviewed 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/m-lab/reboot-service/6#-:-Lc1cJ6o7utHeAlLpam4:bnfp4nl)* status: :shipit: complete! 1 of 1 approvals obtained

**Pull request merged**
